ledger/eval_test.go

```diff
@@ -318,3 +323,225 @@ func TestPrepareAppEvaluators(t *testing.T) {
+	start := time.Now()
+	genesisInitState, addrs, keys := genesis(100000)
+	dbName := fmt.Sprintf("%s.%d", b.Name(), crypto.RandUint64())
+	proto := config.Consensus[genesisInitState.Block.CurrentProtocol]
+	proto.MaxTxnBytesPerBlock = 1000000000 // very big, no limit
+	config.Consensus[protocol.ConsensusVersion(dbName)] = proto
+	genesisInitState.Block.CurrentProtocol = protocol.ConsensusVersion(dbName)
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = true
+	l, err := OpenLedger(logging.Base(), dbName, inMem, genesisInitState, cfg)
+	require.NoError(b, err)
+	defer testLedgerCleanup(l, dbName, inMem)
+
+	dbName2 := dbName + "_2"
+	l2, err := OpenLedger(logging.Base(), dbName2, inMem, genesisInitState, cfg)
+	require.NoError(b, err)
+	defer testLedgerCleanup(l2, dbName2, inMem)
+
+	setupDone := time.Now()
+	setupTime := setupDone.Sub(start)
+	b.Logf("BenchmarkBlockEvaluator setup time %s", setupTime.String())
+
+	bepprof := os.Getenv("BLOCK_EVAL_PPROF")
+	if len(bepprof) > 0 {
+		profpath := dbName + "_cpuprof"
+		profout, err := os.Create(profpath)
+		if err != nil {
+			b.Fatal(err)
+			return
+		}
+		b.Logf("%s: cpu profile for b.N=%d", profpath, b.N)
+		pprof.StartCPUProfile(profout)
+		defer func() {
+			pprof.StopCPUProfile()
+			profout.Close()
+		}()
+	}
+
+	// test speed of block building
+	newBlock := bookkeeping.MakeBlock(genesisInitState.Block.BlockHeader)
+	bev, err := l.StartEvaluator(newBlock.BlockHeader, 0)
+	require.NoError(b, err)
+
+	genHash := genesisInitState.Block.BlockHeader.GenesisHash
+
+	numTxns := 50000
+
+	for i := 0; i < numTxns; i++ {
+		sender := i % len(addrs)
+		receiver := (i + 1) % len(addrs)
+		txn := transactions.Transaction{
+			Type: protocol.PaymentTx,
+			Header: transactions.Header{
+				Sender:      addrs[sender],
+				Fee:         minFee,
+				FirstValid:  newBlock.Round(),
+				LastValid:   newBlock.Round(),
+				GenesisHash: genHash,
+			},
+			PaymentTxnFields: transactions.PaymentTxnFields{
+				Receiver: addrs[receiver],
+				Amount:   basics.MicroAlgos{Raw: 100},
+			},
+		}
+		st := txn.Sign(keys[sender])
+		err = bev.Transaction(st, transactions.ApplyData{})
+		require.NoError(b, err)
+	}
+
+	validatedBlock, err := bev.GenerateBlock()
+	require.NoError(b, err)
+
+	blockBuildDone := time.Now()
+	blockBuildTime := blockBuildDone.Sub(setupDone)
+	b.ReportMetric(float64(blockBuildTime)/float64(numTxns), "ns/block_build_tx")
+
+	l.AddValidatedBlock(*validatedBlock, agreement.Certificate{})
+
+	avbDone := time.Now()
+	avbTime := avbDone.Sub(blockBuildDone)
+	b.ReportMetric(float64(avbTime)/float64(numTxns), "ns/AddValidatedBlock_tx")
+
+	// test speed of block validation
+	// This should be the same as the eval line in ledger.go AddBlock()
+	// This is pulled out to isolate eval() time from db ops of AddValidatedBlock()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err = eval(context.Background(), l2, validatedBlock.blk, false, nil, nil)
```

Comment:
it's "almost" right; the fourth parameter should be "true" to have it perform validation.
I think it would be more reliable to just use
```
_, err = l2.Validate(context.Background(), validatedBlock.blk, nil, nil)
```
```

Reply:
on unrelated topic, the `(l *Ledger) Validate` looks like it should be moved to `ledger.go`. but that's not really related to your change.

Reply:
Do we care about comparing how much time crypto validation takes in this test? In my mind that's the part that's parallelizeable and probably not the bottleneck (I'm betting the serial txn accounting will be the blocker). I guess what I'll do in the next hour is make it both ways and see how different the time is.

Reply:
One number that I'd like to get from this experiment is "how long does it takes to perform a validation in the worst case scenario for a 10MB block". We learned from telemetry that it's about 3 seconds. I want to hear from you how much time does it truly takes. This would help us to determine if this is the bottleneck or not.
A subsequent benchmark would be to see if parallelizing some of that (optimistically) could really reduce time validation time. My guess is "yes", but I don't want to say so before having a concrete test.

Reply:
for a 50,000 txn block, 3,766,519,337 ns aka 3.7 seconds (while using sqlite3 in RAM)

Reply:
could you summarize the results in the PR summary ? It would be great to have it documented.
I think that the missing validation case is typical for when the txpool is full and we can't accept more txns.